### PR TITLE
fix: Creation of an Instance with a Capacity Reservation ID

### LINKED ISSUE #277: Creation of an Intance with a Capacity Reservation ID
## Description

When trying to call the module with the specification of a Capacity Reservation ID, the module ends up with the following error:

╷
│ Error: Invalid function argument
│ 
│   on .terraform/modules/terraform-aws-ec2-instance/main.tf line 45, in resource "aws_instance" "this":
│   45:           capacity_reservation_id = lookup(capacity_reservation_target.value, "capacity_reservation_id", null)
│     ├────────────────
│     │ capacity_reservation_target.value is "cr-xxxxxxxx"
│ 
│ Invalid value for "inputMap" parameter: lookup() requires a map as the
│ first argument.
╵

As per line 45 analysis, it looks like the bug is in the following statement:

capacity_reservation_id = lookup(capacity_reservation_target.value, "capacity_reservation_id", null)

it should rather be as follows:

capacity_reservation_id = lookup(capacity_reservation_target, "capacity_reservation_id", null)

In this case, terraform should lookup in the capacity reservation target map (since it is the iterator in the for_each) rather than in its value for the mapping of the attribute capacity reservation id

- [X] ✋ I have searched the open/closed issues and my issue is not listed.

## Versions

- Module version [Required]: 4.1.0

- Terraform version: v1.2.5

- Provider version(s): v4.22.0

## Reproduction Code [Required]

Call the module with following inputs:

module "ec2_capacity_reservation" {
  source = "../../"

  name = "${local.name}-capacity-reservation"

  ami                         = data.aws_ami.amazon_linux.id
  instance_type               = "u-6tb1.56xlarge"
  subnet_id                   = element(module.vpc.private_subnets, 0)
  vpc_security_group_ids      = [module.security_group.security_group_id]
  associate_public_ip_address = true

  capacity_reservation_specification = {
      capacity_reservation_target = {
        capacity_reservation_id = "cr-00000000000000000"
      }
    }

  tags = local.tags
}

 No
 Yes

## Expected behavior
TF plan should go through and the module should create an EC2 Instance which consumes the capacity reservation.

## Actual behavior

TF plan fails as it is unable to lookup for the capacity reservation id

### PR DESCRIPTION
**Pull request by schniber**
## Description
When trying to call the module with the specification of a Capacity Reservation ID, the module ends up with the following error:

╷
│ Error: Invalid function argument
│
│ on .terraform/modules/terraform-aws-ec2-instance/main.tf line 45, in resource "aws_instance" "this":
│ 45: capacity_reservation_id = lookup(capacity_reservation_target.value, "capacity_reservation_id", null)
│ ├────────────────
│ │ capacity_reservation_target.value is "cr-xxxxxxxx"
│
│ Invalid value for "inputMap" parameter: lookup() requires a map as the
│ first argument.
╵

As per line 45 analysis, it looks like the bug is in the following statement:

capacity_reservation_id = lookup(capacity_reservation_target.value, "capacity_reservation_id", null)

it should rather be as follows:

capacity_reservation_id = lookup(capacity_reservation_target, "capacity_reservation_id", null)

In this case, terraform should lookup in the capacity reservation target map (since it is the iterator in the for_each) rather than in its value for the mapping of the attribute capacity reservation id

## Motivation and Context
Fixes #277 

## Breaking Changes
No breaking changes, just fixing a typo in line 45 in the main.tf

## How Has This Been Tested?
- [X] I have updated at least one of the `examples/*` to demonstrate and validate my change(s)
- [ ] I have tested and validated these changes using one or more of the provided `examples/*` projects
This fix is hard to test since it requires to have an ongoing capacity reservation which is released by AWS.
In order to test this, I stopped at the plan stage.
- [x] I have executed `pre-commit run -a` on my pull request

### CODE REVIEW

**Comment by schniber** [comment]
Hello @bryantbiggs,

Would it be possible for your to review this PR ?

Thanks in advance.

Bests.

Keep up the good work !

**Review by bryantbiggs — CHANGES_REQUESTED** [review]
We shouldn't be deleting the README documentation - not sure why our checks are still passing with all this being deleted. Please update when you get a chance

**Comment by schniber** [comment]
@bryantbiggs  thanks for your feedback.

this was due to the fact that I mistakenly removed the end of tfdocs hook in the complete example README:


Now pre-commit hooks run correctly.

File: examples/complete/main.tf [hunk 1]
```diff
@@ -328,3 +328,27 @@ module "ec2_spot_instance" {
...
+
+  ami                         = data.aws_ami.amazon_linux.id
+  instance_type               = "u-6tb1.56xlarge"
+  subnet_id                   = element(module.vpc.private_subnets, 0)
+  vpc_security_group_ids      = [module.security_group.security_group_id]
+  associate_public_ip_address = true
+
+  capacity_reservation_specification = {
+    capacity_reservation_target = {
+      capacity_reservation_id = "cr-00000000000000000"
```
Comment (bryantbiggs) [inline examples/complete/main.tf:349]: we should use an actual reservation here so its deployable - you can see how we do that here https://github.com/terraform-aws-modules/terraform-aws-autoscaling/blob/master/examples/complete/main.tf#L468-L472

and you can add a cap reservation resource like https://github.com/terraform-aws-modules/terraform-aws-autoscaling/blob/master/examples/complete/main.tf#L796-L802

File: examples/complete/main.tf [hunk 2]
```diff
@@ -328,3 +336,27 @@ module "ec2_spot_instance" {
...
+# EC2 Module - Capacity Reservation
+################################################################################
+
+module "ec2_capacity_reservation" {
+  source = "../../"
+
+  name = "${local.name}-capacity-reservation"
+
+  ami                         = data.aws_ami.amazon_linux.id
+  instance_type               = "m6i.2xlarge"
```
Comment (bryantbiggs) [inline examples/complete/main.tf:350]:
```suggestion
  instance_type               = "t3.micro"
```

**Review by bryantbiggs — APPROVED** [review]
Thank you @schniber !

**Comment by antonbabenko** [comment]
This PR is included in version 4.1.1 :tada:

**Pull request merged**